superopt/src/rewriter.rs

```diff
@@ -120,3 +160,100 @@ impl<G: GraphLike + Clone> CausalRewriter<G> {
         }
     }
 }
+
+#[cfg(test)]
+pub(crate) mod test {
+    use crate::cost::{CostMetric, TwoQubitGateCount};
+    use crate::rewrite_sets::test::rewrite_set_2qb_lc;
+
+    use super::*;
+    use quizx::json::decode_graph;
+    use quizx::vec_graph::Graph;
+    use rstest::{fixture, rstest};
+
+    /// Makes a simple graph, with 2 inputs and 2 outputs, and causal flow.
+    #[fixture]
+    pub(crate) fn small_graph() -> Graph {
+        let mut g = Graph::new();
+        let vs = vec![
+            g.add_vertex(VType::B),
+            g.add_vertex(VType::B),
+            g.add_vertex(VType::Z),
+            g.add_vertex(VType::Z),
+            g.add_vertex(VType::Z),
+            g.add_vertex(VType::Z),
+            g.add_vertex(VType::Z),
+            g.add_vertex(VType::Z),
+            g.add_vertex(VType::Z),
+            g.add_vertex(VType::Z),
+            g.add_vertex(VType::B),
+            g.add_vertex(VType::B),
+        ];
+
+        g.set_inputs(vec![vs[0], vs[1]]);
+        g.set_outputs(vec![vs[10], vs[11]]);
+
+        g.add_edge_with_type(vs[0], vs[2], EType::N);
+        g.add_edge_with_type(vs[1], vs[3], EType::N);
+
+        g.add_edge_with_type(vs[2], vs[4], EType::H);
+        g.add_edge_with_type(vs[3], vs[5], EType::H);
+        g.add_edge_with_type(vs[2], vs[3], EType::H);
+
+        g.add_edge_with_type(vs[4], vs[6], EType::H);
+        g.add_edge_with_type(vs[5], vs[7], EType::H);
+
+        g.add_edge_with_type(vs[6], vs[8], EType::H);
+        g.add_edge_with_type(vs[7], vs[9], EType::H);
+        g.add_edge_with_type(vs[6], vs[7], EType::H);
+
+        g.add_edge_with_type(vs[8], vs[10], EType::N);
+        g.add_edge_with_type(vs[9], vs[11], EType::N);
+
+        g
+    }
+
+    #[fixture]
+    pub(crate) fn json_simple_graph() -> Graph {
+        const SIMPLE_GRAPH_JSON: &str = include_str!("../../test_files/simple-graph.json");
+        decode_graph(SIMPLE_GRAPH_JSON).unwrap()
+    }
+
+    #[fixture]
+    pub(crate) fn compiled_rewriter() -> CausalRewriter<Graph> {
+        let rw_set = rewrite_set_2qb_lc();
+        CausalRewriter::from_rewrite_rules(rw_set)
+    }
+
+    #[fixture]
+    pub(crate) fn pre_compiled_rewriter() -> CausalRewriter<Graph> {
+        const REWRITE_2QB_LC: &[u8] = include_bytes!("../../test_files/rewrites-2qb-lc.rwr");
+        rmp_serde::from_slice(REWRITE_2QB_LC).unwrap()
+    }
+
+    #[rstest]
+    #[case::small_compiled(small_graph(), compiled_rewriter())]
+    #[case::json_compiled(json_simple_graph(), compiled_rewriter())]
+    #[case::small_precompiled(small_graph(), pre_compiled_rewriter())]
+    #[case::json_precompiled(json_simple_graph(), pre_compiled_rewriter())]
```

Comment:
This last case fails.

It's using the extracted LHS graph as the base graph, and the compiled set of rewrites.

If we compile the rewrites instead and use that, it doesn't panic >.<

Reply:
<details>
  <summary> Backtrace </summary>
  
```
thread 'rewriter::test::test_match_apply::case_4_json_precompiled' panicked at /home/aborgna/src/quizx/quizx/src/graph.rs:463:34:
Edge not found
stack backtrace:
   0: rust_begin_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
   2: core::panicking::panic_display
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:178:5
   3: core::panicking::panic_str
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:152:5
   4: core::option::expect_failed
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/option.rs:1985:5
   5: core::option::Option<T>::expect
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/option.rs:894:21
   6: quizx::graph::GraphLike::edge_type
             at /home/aborgna/src/quizx/quizx/src/graph.rs:463:9
   7: quizx::portmatching::pattern::CausalPattern<G>::causal_path_to_line::{{closure}}
             at /home/aborgna/src/quizx/quizx/src/portmatching/pattern.rs:188:50
   8: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:305:13
   9: core::option::Option<T>::map
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/option.rs:1072:29
  10: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/map.rs:103:26
  11: <core::iter::adapters::chain::Chain<A,B> as core::iter::traits::iterator::Iterator>::next::{{closure}}
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/chain.rs:51:67
  12: core::option::Option<T>::or_else
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/option.rs:1511:21
  13: <core::iter::adapters::chain::Chain<A,B> as core::iter::traits::iterator::Iterator>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/chain.rs:51:9
  14: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/vec/spec_from_iter_nested.rs:26:32
  15: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/vec/spec_from_iter.rs:33:9
  16: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/vec/mod.rs:2753:9
  17: core::iter::traits::iterator::Iterator::collect
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/traits/iterator.rs:2054:9
  18: quizx::portmatching::pattern::CausalPattern<G>::causal_path_to_line
             at /home/aborgna/src/quizx/quizx/src/portmatching/pattern.rs:189:16
  19: quizx::portmatching::pattern::CausalPattern<G>::to_lines
             at /home/aborgna/src/quizx/quizx/src/portmatching/pattern.rs:148:24
  20: quizx::portmatching::pattern::CausalPattern<G>::edges
             at /home/aborgna/src/quizx/quizx/src/portmatching/pattern.rs:72:9
  21: quizx::portmatching::matcher::CausalMatcher<G>::get_pattern_matches
             at /home/aborgna/src/quizx/quizx/src/portmatching/matcher.rs:101:58
  22: quizx::portmatching::matcher::CausalMatcher<G>::find_rooted_matches::{{closure}}
             at /home/aborgna/src/quizx/quizx/src/portmatching/matcher.rs:61:41
  23: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:305:13
  24: core::option::Option<T>::map
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/option.rs:1072:29
  25: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/map.rs:103:26
  26: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
  27: core::iter::adapters::fuse::and_then_or_clear
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/fuse.rs:423:13
  28: <core::iter::adapters::fuse::Fuse<I> as core::iter::adapters::fuse::FuseImpl<I>>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/fuse.rs:272:9
  29: <core::iter::adapters::fuse::Fuse<I> as core::iter::traits::iterator::Iterator>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/fuse.rs:43:9
  30: <core::iter::adapters::flatten::FlattenCompat<I,U> as core::iter::traits::iterator::Iterator>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/flatten.rs:480:29
  31: <core::iter::adapters::flatten::FlatMap<I,U,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/flatten.rs:52:9
  32: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
  33: core::iter::adapters::flatten::and_then_or_clear
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/flatten.rs:709:13
  34: <core::iter::adapters::flatten::FlattenCompat<I,U> as core::iter::traits::iterator::Iterator>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/flatten.rs:477:36
  35: <core::iter::adapters::flatten::FlatMap<I,U,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/flatten.rs:52:9
  36: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/map.rs:103:9
  37: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
  38: core::iter::adapters::fuse::and_then_or_clear
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/fuse.rs:423:13
  39: <core::iter::adapters::fuse::Fuse<I> as core::iter::adapters::fuse::FuseImpl<I>>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/fuse.rs:272:9
  40: <core::iter::adapters::fuse::Fuse<I> as core::iter::traits::iterator::Iterator>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/fuse.rs:43:9
  41: <core::iter::adapters::flatten::FlattenCompat<I,U> as core::iter::traits::iterator::Iterator>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/flatten.rs:480:29
  42: <core::iter::adapters::flatten::FlatMap<I,U,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/adapters/flatten.rs:52:9
  43: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/vec/spec_from_iter_nested.rs:26:32
  44: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/vec/spec_from_iter.rs:33:9
  45: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/vec/mod.rs:2753:9
  46: core::iter::traits::iterator::Iterator::collect
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/iter/traits/iterator.rs:2054:9
  47: <quizx_superopt::rewriter::CausalRewriter<G> as quizx_superopt::rewriter::Rewriter>::get_rewrites
             at ./src/rewriter.rs:64:9
  48: quizx_superopt::rewriter::test::test_match_apply
             at ./src/rewriter.rs:246:24
  49: quizx_superopt::rewriter::test::test_match_apply::case_4_json_precompiled
             at ./src/rewriter.rs:234:5
  50: quizx_superopt::rewriter::test::test_match_apply::case_4_json_precompiled::{{closure}}
             at ./src/rewriter.rs:242:10
  51: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
  52: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
test rewriter::test::test_match_apply::case_4_json_precompiled ... FAILED

failures:

failures:
    rewriter::test::test_match_apply::case_4_json_precompiled

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 6 filtered out; finished in 0.09s

error: test failed, to rerun pass `-p quizx-superopt --lib`
```

</details>